### ISSUE TITLE
The wilds of host

### ISSUE THREAD

**Issue by tebaly** [deployphp/deployer#1598]
| Q                 | A
| ----------------- | ---
| Issue Type        | Bug, Question, Feature Request
| Deployer Version  | 6.1
| Local Machine OS  | Linux
| Remote Machine OS | Linux

### Description 
It's all too complicated.

![hasts](https://user-images.githubusercontent.com/9466814/39141238-dca7e9fe-472f-11e8-9340-a48160249b3a.png)

### Content of `deploy.php`
```php
/*
It`s all in one 
- Name of Host
- Hostname for SSH connection
- Name of setup
- Name of stage
*/ 
host('site.net')
    ->user('root');
```

```php 
host('site.net') // - Name of setup - Name of Host -
    ->hostname(167.67.76.7)  // - Hostname for SSH connection
    ->user('root') 
    ->stage('alpha'); // - Name of stage
```

# In docs say

> Often you have only one server for prod and beta stages. You can easily configure them [One host for a few stages](https://deployer.org/docs/hosts#one-host-for-a-few-stages)

```
host('production') // <- What this ?
    ->hostname('domain.com')
    ->set('deploy_path', '~/domain.com');

host('beta') //  <- What this ?
    ->hostname('domain.com')
    ->set('deploy_path', '~/beta.domain.com'); 
```
What this - name of stage now. But where i specify list of hosts? 
**Maybe it's all too complicated?**

---php
 // host fo ssh connection -> name of hosts(list of sites host) -> Stage -> setup name
host(???) 
    ->user('root');

 // ~~host for ssh connection~~ -> name of hosts(list of sites host) -> ~~Stage~~ -> setup name
host(???)
    ->hostname(167.67.76.7)  // - Hostname for SSH connection
    ->user('root') 
    ->stage('alpha'); // - Name of stage

 // ~~host for ssh connection~~ -> ~~name of hosts(list of sites host)~~ -> ~~Stage~~ -> **setup name**
host(???)
    **->hosts('site.net', 'site.me')**
    ->hostname(167.67.76.7)  // - Hostname for SSH connection
    ->user('root') 
    **->stage('alpha')**; // - Name of stage

---

# Handle setup (full)

```php
host('my-setup-name')
    ->hosts('domain.com', 'domain.net', 'domain.biz')
    ->user('name')
    ->hostname('167.67.76.7')
    ->port(22)
    ->configFile('~/.ssh/config')
    ->identityFile('~/.ssh/id_rsa')
    ->forwardAgent(true)
    ->multiplexing(true)
    ->addSshOption('UserKnownHostsFile', '/dev/null')
    ->addSshOption('StrictHostKeyChecking', 'no')
    ->stage('dev');
```
Then root function can be called ~~server~~ -> ~~host~~ -> **setup**

**Comment by antonmedv** [comment]
> Maybe it's all too complicated?

Agree. I'm planning to rework all that system in next Deployer v7.

To be able to do something like this:

```php
stage('production', function () {
    host('name@domain.com');
    host('name@domain.net');
    set('deploy_path', '~');
});
```

**Comment by tebaly** [comment]
If three domains('domain.com', 'domain.net', 'domain.biz')[virtual hosts] in one server(host '1.1.1.1') AND this tree domains in another server(host '2.2.2.2'). Connect to the physical server only by IP address. Name of  virtual hosts used in **deploy_path** like _/var/www/{{hostname}}_ 
* in the new version Deployer v7 it will be possible?

One setup(stage) for multiple nodes of one server with same virtual hosts. 
**Now it might look like this**

```php
setup('my-setup-name')
    ->hosts('domain.com', 'domain.net', 'domain.biz') // list of virtual hosts
    ->user('name')
    ->servers('167.67.76.7', '2.2.2.2', '1.1.1.1') // list of physical server
    ->port(22)
    ->configFile('~/.ssh/config')
    ->identityFile('~/.ssh/id_rsa')
    ->forwardAgent(true)
    ->multiplexing(true)
    ->addSshOption('UserKnownHostsFile', '/dev/null')
    ->addSshOption('StrictHostKeyChecking', 'no')
    ->stage('dev');
```

---
Such a process is expected. scheme

```php
foreach(['167.67.76.7', '2.2.2.2', '1.1.1.1'] as $server) {
    foreach(['domain.com', 'domain.net', 'domain.biz'] as $hostname) {
        set('deploy_path', '/var/www/{{hostname}}'); 
        ...
    }
}
```

**Comment by antonmedv** [comment]
No, i think this is complicated. Instead one host corresponds to one physical host. Or you can use host aliases (`host('name@domain.com/myprojesct')`)

**Comment by tebaly** [comment]
**Multiple hosts** availeble https://deployer.org/docs/hosts#multiple-hosts in docs example 

```php
host('110.164.16.59', '110.164.16.34', '110.164.16.50', ...)
    ->stage('production')
```
There is no possibility to name web-hosts. 
Connection occurs to '110.164.16.59', '110.164.16.34', '110.164.16.50', ...

---

What is happening now?

```php
host('domain.com', 'domain.net', 'domain.biz')
    ->hostname('110.164.16.59')
    ->stage('production')
```
Connection occurs to '110.164.16.59' **only** but  on variable {{hostname}} contains 'domain.com' next 'domain.net' and next 'domain.biz'

it's comfortable, but maybe not normally. 
because this can not be reproduced in inventory file

**Comment by antonmedv** [comment]
In inventory file you can use `<<` yaml extension. Also I'm thinking about removing `->hostname`

**Comment by tebaly** [comment]
> Also I'm thinking about removing ->hostname

Ok.. Bare server. I prepare it to move Web site 'domain.com' from another server. dns to new server is not yet configured. How i can configure deploy to new server without _"hostname"_ parameter?
* i need specify IP addres to server connect
* i need specify hostname 'domain.com' for deploy process
    * like this set('deploy_path', '/var/www/{{**hostname**}}'); 

PS: i have many servers and servers with the same settings. The difference is only in the name of the virtual host and the server address

**Comment by tebaly** [comment]
Another task use variable _hostname_ in 6+ ver

**Comment by antonmedv** [comment]
Fixed in #1904 
Now deployer only have one notion of hosts. Think of it as database. Hosts database.
```
host('domain.com', 'domain.net', 'domain.biz')
    ->hostname('110.164.16.59')
    ->set('deploy_path', '~/{{host_alias}}')
    ->roles('prod');
```
Will converts to table of:
| host_alaias | hostname | deploy_path | roles | ... |
|------------|-----------|-------------|------|---|
| domain.com| 110.164.16.59|~/domain.com| prod|...|
